File: android/src/main/kotlin/com/wembleystudios/sociallogin/FacebookHandler.kt
package com.wembleystudios.sociallogin

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import com.facebook.*
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import org.json.JSONObject

/**
 * Created by Álvaro Blanco Cabrero on 09/11/2018.
 * android.
 */
class FacebookHandler(context: Context, socialConfigOwner: SocialConfigOwner) {

    private val callbackManager: CallbackManager by lazy {
        CallbackManager.Factory.create()
    }

    private var onLoginResult: ((SocialUser?, Throwable?) -> Unit)? = null

    private val callback: FacebookCallback<LoginResult> by lazy {
        object : FacebookCallback<LoginResult> {
            override fun onSuccess(result: LoginResult) {
                fetchUserProfile(result.accessToken) { socialUser, throwable ->
                    onLoginResult?.invoke(socialUser, throwable)
                }
            }

            override fun onCancel() {
                onLoginResult?.invoke(null, null)
            }

            override fun onError(error: FacebookException) {
                onLoginResult?.invoke(null, error)
            }
        }
    }

    init {
        FacebookSdk.setApplicationId(socialConfigOwner.socialConfig.facebookAppId)
        FacebookSdk.sdkInitialize(context.applicationContext)
        LoginManager.getInstance().registerCallback(callbackManager, callback)
    }

    fun getCurrentUserProfile(callback: (SocialUser?, Throwable?) -> Unit) {
        AccessToken.getCurrentAccessToken()?.let {
            fetchUserProfile(it, callback)
        } ?: run {
            callback(null, NO_USER_LOGGED_IN_ERROR)
        }
    }

    fun logInWithReadPermissions(
        activity: Activity,
        permissions: List<String>,
        callback: (SocialUser?, Throwable?) -> Unit
    ) {
        if (AccessToken.getCurrentAccessToken()?.permissions?.containsAll(permissions) == true) {
            fetchUserProfile(AccessToken.getCurrentAccessToken(), callback)
        } else {
            onLoginResult = callback
            LoginManager.getInstance().logInWithReadPermissions(activity, permissions)
        }
    }

    fun logOut() = LoginManager.getInstance().logOut()

    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) =
        callbackManager.onActivityResult(requestCode, resultCode, data)

    private fun fetchUserProfile(
        accessToken: AccessToken,
        callback: (SocialUser?, Throwable?) -> Unit
    ) {
        val request =
            GraphRequest.newMeRequest(accessToken) { jsonObject, response ->
                response.error?.let {
                    callback(null, response.error.exception)
                } ?: run {
                    callback(
                        SocialUser(
                            jsonObject.getString(ME_REQUEST_RESPONSE_ID),
                            jsonObject.getString(ME_REQUEST_RESPONSE_EMAIL),
                            jsonObject.getString(ME_REQUEST_RESPONSE_NAME),
                            getProfilePicture(jsonObject),
                            mapOf(Constants.FACEBOOK_TOKEN to accessToken.token)
                        ),
                        null
                    )
                }
            }
        request.parameters = Bundle().apply {
            putString(ME_REQUEST_PARAM_FIELDS_KEY, ME_REQUEST_PARAM_FIELDS_VALUE)
        }
        request.executeAsync()
    }

    private fun getProfilePicture(jsonObject: JSONObject): String? {
        val pictureJson = jsonObject.getJSONObject(ME_REQUEST_RESPONSE_PICTURE)
        val data = pictureJson?.getJSONObject(ME_REQUEST_RESPONSE_DATA)
        return data?.getString(ME_REQUEST_RESPONSE_URL)
    }

    companion object {
        private const val ME_REQUEST_PARAM_FIELDS_KEY = "fields"
        private const val ME_REQUEST_PARAM_FIELDS_VALUE = "email, name, picture"

        private const val ME_REQUEST_RESPONSE_ID = "id"
        private const val ME_REQUEST_RESPONSE_EMAIL = "email"
        private const val ME_REQUEST_RESPONSE_NAME = "name"
        private const val ME_REQUEST_RESPONSE_PICTURE = "picture"
        private const val ME_REQUEST_RESPONSE_DATA = "data"
        private const val ME_REQUEST_RESPONSE_URL = "url"

        private val NO_USER_LOGGED_IN_ERROR = FacebookException("No user logged in")
    }
}

File: ios/Classes/Handlers/FacebookHandler.swift
//
//  FacebookHandler.swift
//  Runner
//
//  Created by Luis Paez Gonzalez on 13/11/2018.
//  Copyright © 2018 The Chromium Authors. All rights reserved.
//

import UIKit

import Flutter
import FBSDKCoreKit
import FBSDKLoginKit

class FacebookHandler {
    
    // MARK: Constants
    private static let ME_REQUEST_PARAM_FIELDS_KEY = "fields"
    private static let ME_REQUEST_PARAM_FIELDS_VALUE = "email, name, picture"
    private static let ME_REQUEST_RESPONSE_ID = "id"
    private static let ME_REQUEST_RESPONSE_EMAIL = "email"
    private static let ME_REQUEST_RESPONSE_NAME = "name"
    private static let ME_REQUEST_RESPONSE_PICTURE = "picture"
    private static let ME_REQUEST_RESPONSE_DATA = "data"
    private static let ME_REQUEST_RESPONSE_URL = "url"
    
    
    // MARK: Methods
    class func configuration(facebookAppID: String) {
        FBSDKSettings.setAppID(facebookAppID)
    }
        
    class func logInFacebookWithPermissions(permissions: [String], result: @escaping FlutterResult) {
        
        let sequence = Set(permissions)
        
        if FBSDKAccessToken.current() != nil {
            
        } else {
            let login = FBSDKLoginManager.init()
            login.logIn(withReadPermissions: permissions, from: UIApplication.shared.keyWindow?.rootViewController) { (fbresult, error) in
                
                if error != nil {
                    result (false)
                    return
                }
                
                if let bool = fbresult?.isCancelled, bool == true {
                    result (false)
                    return
                }
                
                if (fbresult?.token) != nil {
                    result (true)
                    return
                }
                
                result(false)
            }
        }
    }
    
    class func logOutFacebook() {
        FBSDKLoginManager().logOut()
        FBSDKAccessToken.setCurrent(nil)
        FBSDKProfile.setCurrent(nil)
    }
}
